### PR TITLE
fix: clearPolicy should save policy

### PR DESCRIPTION
**Pull request by zhmushan**

### CODE REVIEW

**Comment by hsluoyz** [comment]
@zhmushan Why `clearPolicy should save policy `?

**Comment by zhmushan** [comment]
Need discussion with @nodece

**Comment by nodece** [comment]
@hsluoyz `clearPolicy` should be look like `addPolicy` and `removePolicy`, we need to sync the model data to storage by adapter.

we just call model.clearPolicy if you need to clear the memory policy.

**Comment by hsluoyz** [comment]
If `AutoSave` is enabled, then we can do `clearPolicy should save policy `

**Comment by nodece** [comment]
@hsluoyz yes. I realized the issue.

File: src/coreEnforcer.ts [hunk 1]
```diff
@@ -138,8 +138,9 @@ export class CoreEnforcer {
   /**
    * clearPolicy clears all policy.
    */
-  public clearPolicy(): void {
+  public async clearPolicy(): Promise<void> {
     this.model.clearPolicy();
+    await this.savePolicy();
```
Comment (nodece) [inline src/coreEnforcer.ts:143]: you should add check `autoSave` in here.

File: src/syncedEnforcer.ts [hunk 2]
```diff
@@ -42,10 +42,13 @@ export class SyncedEnforcer extends Enforcer {
   /**
    * clearPolicy clears all policy.
    */
-  public clearPolicy(): void {
+  public async clearPolicy(): Promise<void> {
     this.lock
       .acquireAsync()
-      .then(() => super.clearPolicy())
+      .then(async () => {
```
Comment (nodece) [inline src/syncedEnforcer.ts:48]: you should add check `autoSave` in here.

**Comment by nodece** [comment]
@zhmushan please fix CI.

**Comment by hsluoyz** [comment]
@nodece please review.

File: src/syncedEnforcer.ts [hunk 3]
```diff
@@ -42,8 +42,8 @@ export class SyncedEnforcer extends Enforcer {
   /**
    * clearPolicy clears all policy.
    */
-  public clearPolicy(): void {
-    this.lock
+  public async clearPolicy(): Promise<void> {
+    await this.lock
       .acquireAsync()
       .then(() => super.clearPolicy())
```
Comment (nodece) [inline src/syncedEnforcer.ts:48]: there is a deadlock here.

Comment (zhmushan) [reply src/syncedEnforcer.ts]: done

File: test/rbacAPI.test.ts [hunk 4]
```diff
@@ -166,7 +166,8 @@ test('test getImplicitPermissionsForUser', async () => {
   expect(await e.getImplicitUsersForPermission('data2', 'read')).toEqual(['alice']);
   expect(await e.getImplicitUsersForPermission('data2', 'write')).toEqual(['alice', 'bob']);
 
-  e.clearPolicy();
+  e.enableAutoSave(false);
+  await e.clearPolicy();
```
Comment (nodece) [inline test/rbacAPI.test.ts:171]: You lost `e.enableAutoSave(true);`

**Comment by nodece** [comment]
@hsluoyz @GopherJ could you review the PR? because JavaScript js single thread, so we remove method of SyncedEnforcer.

**Comment by GopherJ** [comment]
I'm not sure why we removed SyncedEnforcer, javascript is single threaded, but still we can have race.

For example:

```javascript
redis.Get(key, (reply) => {
  redis.Set(key, 2 * reply)
})
```

**Comment by GopherJ** [comment]
See also: https://github.com/casbin/node-casbin/issues/115

**Comment by nodece** [comment]
@GopherJ  You are right, I realized my issues.

**Comment by nodece** [comment]
@zhmushan thanks for your PR.  I will close the PR, because it is incorrect implement.

**Pull request closed without merging**